kernel/src/scan/state.rs

```diff
@@ -1,57 +1,139 @@
+///        data.as_ref(),
+///        vector,
+///        context,
+///        my_callback,
+///     )?;
+/// }
+/// ```
+pub fn visit_scan_files<T>(
+    data: &dyn EngineData,
+    selection_vector: Vec<bool>,
+    context: T,
+    callback: fn(
+        context: &mut T,
+        path: &str,
+        size: i64,
+        dv_info: DvInfo,
+        partition_values: HashMap<String, String>,
+    ) -> (),
+) -> DeltaResult<T> {
+    let mut visitor = ScanFileVisitor {
+        callback,
+        selection_vector,
+        context,
+    };
+    data.extract(super::file_stream::SCAN_ROW_SCHEMA.clone(), &mut visitor)?;
+    Ok(visitor.context)
 }
 
-// will probably actually use this one
+// add some visitor magic for clients
+struct ScanFileVisitor<T> {
+    callback: fn(&mut T, &str, i64, DvInfo, HashMap<String, String>) -> (),
+    selection_vector: Vec<bool>,
+    context: T,
+}
 
-/// State that doesn't change beween scans
-#[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct GlobalScanState {
-    pub table_root: String,
-    pub partition_columns: Vec<String>,
-    pub logical_schema: Schema,
-    pub read_schema: Schema,
+impl<T> DataVisitor for ScanFileVisitor<T> {
+    fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
+        for row_index in 0..row_count {
+            if !self.selection_vector[row_index] {
+                // skip skipped rows
+                continue;
+            }
+            // Since path column is required, use it to detect presence of an Add action
+            if let Some(path) = getters[0].get_opt(row_index, "scanFile.path")? {
+                let size = getters[1].get(row_index, "scanFile.size")?;
+                let deletion_vector = if let Some(storage_type) =
+                    getters[3].get_opt(row_index, "scanFile.deletionVector.storageType")?
+                {
+                    // there is a storageType, so the whole DV must be there
+                    let path_or_inline_dv: String =
+                        getters[4].get(row_index, "scanFile.deletionVector.pathOrInlineDv")?;
+                    let offset: Option<i32> =
+                        getters[5].get_opt(row_index, "scanFile.deletionVector.offset")?;
+                    let size_in_bytes: i32 =
+                        getters[6].get(row_index, "scanFile.deletionVector.sizeInBytes")?;
+                    let cardinality: i64 =
+                        getters[7].get(row_index, "scanFile.deletionVector.cardinality")?;
+                    Some(DeletionVectorDescriptor {
+                        storage_type,
+                        path_or_inline_dv,
+                        offset,
+                        size_in_bytes,
+                        cardinality,
+                    })
+                } else {
+                    None
+                };
+                let dv_info = DvInfo { deletion_vector };
+                let partition_values: HashMap<_, _> =
```

Comment:
I'm slightly surprised the type annotation is needed, given that passing it as a callback arg already constrains it to be a `HashMap<String, String>`?

Reply:
indeed. copy pasted code